tmpl/wiki/tutorial-lwt.md

```diff
@@ -74,508 +111,327 @@ Assuming you are building for POSIX using native kernel sockets, compile the app
+##User-defined threads
 
-Here are the needed functions from the `Lwt_mvar` module:
+You can create a thread that sleeps until you explicitly make it return a result with `Lwt.wait`,
+which returns a thread and a _waker_:
 
 ```
-  (* type of a mailbox variable *)
-  type 'a t
-
-  (* creates a new empty mailbox variable *)
-  val create_empty : unit -> 'a t
-
-  (* puts a value into a mailbox variable, and blocks if it is full *)
-  val put : 'a t -> 'a -> unit Lwt.t
-
-  (* will take any available value and block if the mailbox is empty *)
-  val take : 'a t -> 'a Lwt.t
+let invoke_remote msg =
+  let t, waker = Lwt.wait () in
+  let id = new_id () in
+  on_response_to id (fun resp -> Lwt.wakeup waker resp);
+  send_request id msg;
+  t
 ```
 
-###Challenge
+This is mainly useful when interacting with external processes (as in this example), or libraries that don't support Lwt directly,
 
-Write a small set of functions to help do pipeline parallelism. The interface to be implemented is the following (names should give away the appropriate semantic):
 
-```
-  val map: ('a -> 'b Lwt.t) -> 'a Lwt_mvar.t -> 'b Lwt_mvar.t
-  val split : ('a * 'b) Lwt_mvar.t -> 'a Lwt_mvar.t * 'b Lwt_mvar.t
-  val filter: ('a -> bool Lwt.t) -> 'a Lwt_mvar.t -> 'a Lwt_mvar.t
-```
+##Cancelling
 
-### Solution
+In order to cancel a thread, the function `cancel` (provided by the module Lwt) is needed. It has type `'a t -> unit` and does exactly what it says (except on certain complicated cases that are not in the scope of this tutorial). A simple timeout function that cancels a thread after a given number of seconds can be written easily:
 
 ```
-let map f m_in =
-  let m_out = Lwt_mvar.create_empty () in
-  let rec aux () =
-    Lwt_mvar.(
-      take m_in   >>=
-      f           >>= fun v ->
-      put m_out v >>
-      aux ()
-    )
-  in
-  let t = aux () in
-  m_out
-
-let split m_ab =
-  let m_a, m_b = Lwt_mvar.(create_empty (), create_empty ()) in
-  let rec aux () =
-    Lwt_mvar.take m_ab >>= fun (a, b) ->
-    join [
-      Lwt_mvar.put m_a a;
-      Lwt_mvar.put m_b b;
-    ] >> aux ()
-  in
-  let t = aux () in
-  (m_a, m_b)
-
-let filter f m_a =
-  let m_out = Lwt_mvar.create_empty () in
-  let rec aux () =
-    Lwt_mvar.take m_a >>= fun a ->
-    f a >>= function
-      | true -> Lwt_mvar.put m_out a >> aux ()
-      | false -> aux ()
-  in
-  let t = aux () in
-  m_out
+  (* In this example and all those afterwards, we consider Lwt and OS to be
+     opened *)
+  let timeout delay t =
+    Time.sleep delay >|= fun () -> cancel t
 ```
 
-This code can also be found in `lwt/src/mvar_unikernels.ml` in the `mirage-skeleton` repository.
+Note: this example also shows a new convenience operator provided by Lwt. `>|=` ("map") is similar to `>>=` but automatically wraps the result of the function you provide with `return`. It's used here because `cancel` is synchronous (it doesn't return a thread).
```

Comment:
The "Lwt. `>|=`" looks like a failed attempt to write `Lwt.(>|=)` when rendered – perhaps you can change the period to a colon or dash?